### DISCUSSION TITLE
Assertion failure messages

### DISCUSSION THREAD

**Discussion by 0xfded** [thomhurst/TUnit#574] [Q&A]
Is there a way to display more information on an assertion failure?   In particular when you're in a loop it's helpful to know what item or index caused the failure.

In the example below, there's really no way to know from the failure which sector was involved.

```
  foreach(var sector in fmgr.GetTrackSectorList(fde))  
  {
    await Assert.That(fmgr.VTOC.IsSectorFree(sector)).IsFalse();
  }
```

NUnit has something like:
   `Assert.That(fmgr.VTOC.IsSectorFree(sector), Is.False, $"Sector {sector} is marked free.");`
   
PS - the framework is awesome, thanks for sharing it.   It's fast and has so many options.

**Comment by thomhurst** [comment]
Hey @0xfded you just need support for custom messages? I've mostly built this I've just been trying to work out the best way to expose this on the API in combination with chaining without making it too messy. Which might not actually be possible. I'll take a look. If I can't think of a way to improve it I'll just open up a method "With message" that returns the same assertion but with your overridden message. If that would solve your issue?

**Comment by 0xfded** [comment]
Hi @thomhurst , thanks for the quick reply.   I understand wanting to keep the API clean and intuitive.    If I'm the only person that is missing having additional context for failures, maybe I'm not writing good unit tests :smile:.  

I've written a lot of tests that iterate through structures and have found it helpful to be able to easily identify the object that was involved in the failure.

Rather than making a quick fix, perhaps we can leave this open for a bit and see if anyone else needs it or has suggestions on how to handle these situations.   In the meantime, I'll continue to work with the framework, maybe I'm just not approaching the problem correctly.

**Reply by thomhurst** [reply]
Absolutely! Keep experimenting and also let me know of any ideas! Would love to hear them :)

**Comment by maxime-paquatte** [comment]
Hi,

It seems that there is a "Because" method:

` await Assert.That(fmgr.VTOC.IsSectorFree(sector)).IsFalse().Because($"Sector {sector} is marked free.");`

**Comment by eisnstein** [comment]
Hi @thomhurst, I basically have the same issue right now. I also have a loop in a test. When within the loop an assertion fails, I cannot tell for which item. So I was also looking for some method to add a custom error message. The `Because` method did not help.

My loop looks like this:

```
foreach (var drug in _drugs)
{
    var foundDrug = drugs.FirstOrDefault(d => d.name.Contains(drug, StringComparison.OrdinalIgnoreCase));
    // Here I dont know which drug name was not found
    await Assert.That(foundDrug).IsNotNull();
    // Something like this maybe:
    await Assert.That(foundDrug).IsNotNull().Context($"{drug} was not found in _drugs")
}
```

**Reply by thomhurst** [reply]
What issues did you have with the Because method?

**Reply by gerardryaneagersautomotive** [reply]
For my use case, "Because" is "why this assertion should pass" but I want "why this assertion has failed".
Using Because anyway will make the error message not making any sense in English or have messages phrased in negative terms which is hard to follow.

**Comment by gerardryaneagersautomotive** [comment]
IMO this is most needed in situations like `IsSuccessStatusCode()` for HTTP responses, currently it outputs `Content: System.Net.Http.StreamContent` which might be fine in a general case but in some cases that is the most important context for diagnosing a failing test.
